File: 031/031.rb
#! /usr/bin/env ruby

@denom = [200,100,50,20,10,5,2,1]

def count(n, max)
  return 1 if n == 0
  @denom.drop_while{|d| d > max}.map{|d| n - d < 0 ? 0 : count(n - d, d)}.reduce(:+)
end

puts count(ARGV[0].to_i, @denom.sort.reverse.first)

File: 031/031.rs
static denom : &'static [int] = &[200,100,50,20,10,5,2,1];

fn main() {
  println!("{}", count(200, 200));
}

fn count(n : int, max : int) -> int {
  if n == 0{ return 1; }

  denom.iter()
       .filter(|d| {**d <= max})
       .map(|d| { if n - *d < 0 {0}
                  else {count(n - *d, *d)}
                }
           )
       .fold(0,|acc, x| {acc + x})
}
